src/utils/async.js

```diff
@@ -1,42 +1,57 @@
 /* @flow strict-local */
 import { isClientError } from '../api/apiErrors';
 
-/** The time at which we last started a timeout. */
-let lastTimeoutTime = 0;
+/**
+ * Makes a machine that can sleep for a timeout that, until a ceiling is reached,
+ * grows exponentially in duration with the number of sleeps completed, with a
+ * base of 2. The machine should be created before a loop starts, and .wait()
+ * should be called in each iteration of the loop. The machine should not be
+ * re-used after exiting the loop.
+ *
+ * The .numSleepsCompleted() and .timeElapsedSoFar() getters can be used if the
+ * caller needs to implement "give up" logic for a network request by breaking
+ * out of the loop after certain threshold. Note that .timeElapsedSoFar() will
+ * give the total time elapsed since the start of the first sleep, so this will
+ * include the time taken by other tasks like API requests.
+ *
+ * E.g., if firstDuration is 100 and durationCeiling is 10 * 1000 = 10000,
+ * the sequence is
+ *
+ * 100, 200, 400, 800, 1600, 3200, 6400, 10000, 10000, 10000, ...
+ */
+export const makeBackoffMachine = () => {
```

Comment:
I'd suggest defining `type BackoffMachine` explicitly and specifying it as this function's return type.

(Passing objects of only-implicitly-defined type across different contexts like this always feels a bit sketchy to me.)

Reply:
Yep, that makes sense.